cobaya/model.py

```diff
@@ -264,19 +385,38 @@ def logps(self, input_params, return_derived=True, cached=True, make_finite=Fals
                         component.current_logp)  # type: ignore
         if make_finite:
             loglikes = np.nan_to_num(loglikes)
-        if return_derived:
-            # Turn the derived params dict into a list and return
+        if as_dict:
+            loglikes = dict(zip(self.likelihood, loglikes))
+        if return_derived or return_output_params:
             if not compute_success:
-                derived_list = [np.nan] * len(self.output_params)
+                return_params_names = (self.output_params if return_output_params else
+                                       self.parameterization.derived_params)
+                if as_dict:
```

Comment:
Should maybe make self.derived_params list similar to self.out_params for consistency and avoid dict copy in parameterization.derived_params?

Reply:
Done, thanks!